File: python/ClapAPI.py
import requests


class ClapAPI:

    def __init__(self, client_id, client_secret, api=None):
        self._api = api or "https://api.clap.video/api/"
        self._max_photos = 5
        self._client_id = client_id
        self._client_secret = client_secret

        self._headers = None
        self._expires_in = None
        self._access_token = None

        self._generate_access_token()

    def _generate_url(self, endpoint):
        return "{}{}".format(self._api, endpoint)

    def _make_request(self, method, url, data=None, headers=None, error=None):
        if not headers:
            headers = self._headers

        url = self._generate_url(url)

        req = requests.request(method, url, json=data, headers=headers)

        if req.status_code not in [requests.codes.ok, requests.codes.created]:
            try:
                detail = req.json().get('detail')
            except ValueError as e:
                detail = None

            raise Exception(error or "{} Error while making a Clap Request".format(req.status_code), detail or req.text)

        return req.json()

    def _generate_access_token(self):
        req = self._make_request("POST", "oauth",
                                 data={"client_id": self._client_id, "client_secret": self._client_secret},
                                 error="Error while generating Access Token")

        self._access_token = req.get('access_token')
        self._expires_in = req.get('expires_in', 0)
        self._headers = {"Authorization": "Access-Token {}".format(self._access_token)}

    def get_project(self, id):
        return self._make_request("GET", "video/projects/{}".format(id))

    def create_project(self, **kwargs):
        json_fields = kwargs.get('json_fields', {})

        if kwargs.get('webhook_infos'):
            json_fields['webhook_infos'] = kwargs.get('webhook_infos')
            del kwargs['webhook_infos']
        if kwargs.get('realty'):
            json_fields['realty'] = kwargs.get('realty')
            del kwargs['realty']

        photos_list = kwargs.get('photos', [])

        if photos_list and type(photos_list) is list:
            if not json_fields.get('realty'):
                json_fields['realty'] = {}

            i = 1
            for photo in photos_list:
                json_fields['realty']['photo{}'.format(i)] = photo

                if i < self._max_photos:
                    i += 1
                else:
                    break
            del kwargs["photos"]

        kwargs['json_fields'] = json_fields

        return self._make_request("POST", "video/projects", data=kwargs)

    def get_user_token(self, user_id):
        req = self._make_request("POST", "users/{}/get-auth-token".format(user_id))

        return req.get('token')

File: php/ClapAPI.php
<?php

class ClapAPI{

    protected $api = "https://api.clap.video/api/";
    protected $maxPhotos = 5;
    protected $headers = null;
    protected $expires_in = null;
    protected $accessToken = null;

    protected $clientId = "";
    protected $clientSecret = "";

    public function __construct($clientId, $clientSecret, $api = null){
        $this->clientId = $clientId;
        $this->client_secret = $clientSecret;

        $this->api = $api ?? $this->api;

        $this->generateAccessToken();
    }

    protected function generateUrl($endpoint){
        return "{$this->api}{$endpoint}";
    }

    protected function makeRequest($method, $url, $data = null, $headers = null, $error = null){
        $url = $this->generateUrl($url);

        $headers = array('Content-Type: application/json', $this->headers ?? '');

        $data_json = json_encode($data);
        $ch = curl_init($url);
        curl_setopt($ch, CURLOPT_CUSTOMREQUEST, $method);
        curl_setopt($ch, CURLOPT_SSL_VERIFYHOST, 0);
        curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, 0);

        if($method == "POST")
            curl_setopt($ch, CURLOPT_POSTFIELDS, $data_json);


        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
        curl_setopt($ch, CURLOPT_HTTPHEADER, $headers);
        curl_setopt($ch, CURLOPT_TIMEOUT, 5);
        curl_setopt($ch, CURLOPT_CONNECTTIMEOUT, 5);

        $result = json_decode(curl_exec($ch), true);
        curl_close($ch);

        return $result;
    }

    protected function generateAccessToken(){
        $req = $this->makeRequest(
            "POST", "oauth",
            ["client_id" => $this->clientId, "client_secret" => $this->client_secret],
            "Error while generating Access Token"
        );

        $this->accessToken = isset($req["access_token"]) ? $req["access_token"] : null;
        $this->expiresIn = isset($req["expires_in"]) ? $req["expires_in"] : null;
        $this->headers = "Authorization: Access-Token {$this->accessToken}";
    }

    public function getProject($id){
        return $this->makeRequest("GET", "video/projects/{$id}");
    }

    public function createProject($options){
        $json_fields = $options['json_fields'] ?? [];

        if(isset($options['webhook_infos'])){
            $json_fields['webhook_infos'] = $options['webhook_infos'];
            unset($options['webhook_infos']);
        }

        if(isset($options['realty'])){
            $json_fields['realty'] = $options['realty'];
            unset($options['realty']);
        }

        $photos_list = $options['photos'] ?? [];

        if($photos_list){
            if(!isset($json_fields['realty'])){
                $json_fields['realty'] = [];
            }

            foreach($photos_list as $i=>$photo){
                $json_fields['realty']['photo'.($i+1)] = $photo;

                if(($i+1) >= $this->maxPhotos){
                    break;
                }
            }
            unset($options['photos']);
        }

        $options['json_fields'] = $json_fields;

        return $this->makeRequest("POST", "video/projects", $options);
    }

    public function getUserToken($userId){
        $req = $this->makeRequest("POST", "users/{$userId}/get-auth-token");

        return isset($req['token']) ? $req['token'] : null;
    }
}
?>
